internal/rbd/rbd_journal.go

```diff
@@ -137,26 +140,42 @@ func checkSnapExists(ctx context.Context, rbdSnap *rbdSnapshot, cr *util.Credent
 			snapData.ImagePool, rbdSnap.Pool)
 	}
 
+	vol := generateVolFromSnap(rbdSnap)
+
 	// Fetch on-disk image attributes
-	err = updateSnapWithImageInfo(ctx, rbdSnap, cr)
+	err = vol.updateVolWithImageInfo(ctx, cr)
 	if err != nil {
-		if _, ok := err.(ErrSnapNotFound); ok {
-			err = j.UndoReservation(ctx, rbdSnap.JournalPool,
-				rbdSnap.Pool, rbdSnap.RbdSnapName, rbdSnap.RequestName)
-			return false, err
+		if _, ok := err.(ErrImageNotFound); ok {
+			err = undoSnapshotCloning(ctx, parentVol, rbdSnap, vol, cr, false)
 		}
 		return false, err
 	}
 
+	rbdSnap.CreatedAt = vol.CreatedAt
+	rbdSnap.SizeBytes = vol.VolSize
 	// found a snapshot already available, process and return its information
 	rbdSnap.SnapID, err = util.GenerateVolID(ctx, rbdSnap.Monitors, cr, snapData.ImagePoolID, rbdSnap.Pool,
 		rbdSnap.ClusterID, snapUUID, volIDVersion)
 	if err != nil {
 		return false, err
 	}
 
-	klog.V(4).Infof(util.Log(ctx, "found existing snap (%s) with snap name (%s) for request (%s)"),
+	klog.V(4).Infof(util.Log(ctx, "found existing image (%s) with name (%s) for request (%s)"),
 		rbdSnap.SnapID, rbdSnap.RbdSnapName, rbdSnap.RequestName)
+	// Madhu TODO check flatten in progress
+	err = vol.Connect(cr)
+	if err != nil {
+		return false, err
+	}
+	vol.Destroy()
+
+	err = vol.flattenRbdImage(ctx, cr)
```

Comment:
(nit) Flatten does not seem to belong to this function as such, which checks if the journal and hence the image is intact. I would see if we can move the flatten check outside of this code to the caller.